Private abstract functions are not spotted by autocxx

The latest autocxx (4f34e24f1f5) fails to build protobuf-sys:

```
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:138:5: error: field type 'google::protobuf::compiler::MultiFileErrorCollector' is an abstract class
warning:   T value;
warning:     ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:1841:145: note: in instantiation of template class 'rust::MaybeUninit<google::protobuf::compiler::MultiFileErrorCollector>' requested here
warning:   ::google::protobuf::compiler::MultiFileErrorCollector *uninit = reinterpret_cast<::google::protobuf::compiler::MultiFileErrorCollector *>(new ::rust::MaybeUninit<::google::protobuf::compiler::MultiFileErrorCollector>);
warning:                                                                                                                                                 ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-src-06001173b1f36074/out/include/google/protobuf/compiler/importer.h:202:16: note: unimplemented pure virtual method 'AddError' in 'MultiFileErrorCollector'
warning:   virtual void AddError(const std::string& filename, int line, int column,
warning:                ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:138:5: error: field type 'google::protobuf::MessageFactory' is an abstract class
warning:   T value;
warning:     ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-sys-1d46895fccfcab95/out/autocxx-build-dir/cxx/gen0.cxx:1997:107: note: in instantiation of template class 'rust::MaybeUninit<google::protobuf::MessageFactory>' requested here
warning:   ::google::protobuf::MessageFactory *uninit = reinterpret_cast<::google::protobuf::MessageFactory *>(new ::rust::MaybeUninit<::google::protobuf::MessageFactory>);
warning:                                                                                                           ^
warning: /Users/benesch/Sites/materialize/rust-protobuf-native/target/debug/build/protobuf-src-06001173b1f36074/out/include/google/protobuf/message.h:1281:26: note: unimplemented pure virtual method 'GetPrototype' in 'MessageFactory'
warning:   virtual const Message* GetPrototype(const Descriptor* type) = 0;
warning:                          ^
warning: 2 errors generated.

error: failed to run custom build command for `protobuf-sys v0.1.2+3.19.1 (/Users/benesch/Sites/materialize/rust-protobuf-native/protobuf-sys)`
```

Whereas v0.14.0 was able to produce a fair number of bindings: https://docs.rs/protobuf-sys/latest/protobuf_sys/

## Steps to Reproduce the Problem

```shell
git clone https://github.com/MaterializeInc/rust-protobuf-native
patch -p1 <<EOF
diff --git a/protobuf-sys/Cargo.toml b/protobuf-sys/Cargo.toml
index 560d18c..ebdf0a7 100644
--- a/protobuf-sys/Cargo.toml
+++ b/protobuf-sys/Cargo.toml
@@ -26,9 +26,9 @@ version = "0.1.2+3.19.1"
 edition = "2021"
 
 [dependencies]
-autocxx = "0.14.0"
+autocxx = { git = "https://github.com/google/autocxx "}
 cxx = "1.0.54"
 protobuf-src = { path = "../protobuf-src", version = "1.0.0" }
 
 [build-dependencies]
-autocxx-build = "0.14.0"
+autocxx-build = { git = "https://github.com/google/autocxx "}
EOF
cd protobuf-sys
cargo check
```

If this is too large of a reproducer, feel free to kick the can back to me, and I'll try to find the time to minimize it. I figured I'd file early, though, because @adetaylor I thought you might be able to immediately spot the problem based on the error message.

--- a/engine/src/conversion/analysis/abstract_types.rs
+++ b/engine/src/conversion/analysis/abstract_types.rs
@@ -18,7 +18,11 @@ use super::{
     fun::{FnAnalysis, FnKind, FnPhase, MethodKind},
     pod::PodAnalysis,
 };
-use crate::conversion::{api::TypeKind, error_reporter::convert_item_apis, ConvertError};
+use crate::conversion::{
+    api::TypeKind,
+    error_reporter::{convert_apis, convert_item_apis},
+    ConvertError,
+};
 use crate::{conversion::api::Api, types::QualifiedName};
 use std::collections::HashSet;
 
@@ -126,4 +130,28 @@ fn any_missing_from_allowlist(config: &IncludeCppConfig, bases: &HashSet<Qualifi
     bases
         .iter()
         .any(|qn| !config.is_on_allowlist(&qn.to_cpp_name()))
+}
+
+pub(crate) fn discard_ignored_functions(apis: Vec<Api<FnPhase>>) -> Vec<Api<FnPhase>> {
+    // Some APIs can't be generated, e.g. because they're protected.
+    // Now we've finished analyzing abstract types and constructors, we'll
+    // convert them to IgnoredI
+    let mut apis_new = Vec::new();
+    convert_apis(
+        apis,
+        &mut apis_new,
+        |name, fun, analysis, name_for_gc| {
+            analysis.ignore_reason.clone()?;
+            Ok(Box::new(std::iter::once(Api::Function {
+                name,
+                fun,
+                analysis,
+                name_for_gc,
+            })))
+        },
+        Api::struct_unchanged,
+        Api::enum_unchanged,
+        Api::typedef_unchanged,
+    );
+    apis_new
 }

--- a/engine/src/conversion/analysis/fun/mod.rs
+++ b/engine/src/conversion/analysis/fun/mod.rs
@@ -136,11 +136,14 @@ pub(crate) struct FnAnalysis {
     pub(crate) vis: Visibility,
     pub(crate) cpp_wrapper: Option<CppFunction>,
     pub(crate) deps: HashSet<QualifiedName>,
-    /// Protected methods still need to be recorded because we want
+    /// Some methods still need to be recorded because we want
     /// to (a) generate the ability to call superclasses, (b) create
     /// subclass entries for them. But we do not want to have them
     /// be externally callable.
-    pub(crate) generate_code: bool,
+    pub(crate) ignore_reason: Result<(), ConvertErrorWithContext>,
+    /// Whether this can be called by external code. Not so for
+    /// protected methods.
+    pub(crate) externally_callable: bool,
 }
 
 #[derive(Clone)]
@@ -158,6 +161,17 @@ struct ReturnTypeAnalysis {
     conversion: Option<TypeConversionPolicy>,
     was_reference: bool,
     deps: HashSet<QualifiedName>,
+}
+
+impl Default for ReturnTypeAnalysis {
+    fn default() -> Self {
+        Self {
+            rt: parse_quote! {},
+            conversion: Default::default(),
+            was_reference: Default::default(),
+            deps: Default::default(),
+        }
+    }
 }
 
 pub(crate) struct FnPhase;
@@ -178,7 +192,6 @@ pub(crate) struct FnAnalyzer<'a> {
     config: &'a IncludeCppConfig,
     overload_trackers_by_mod: HashMap<Namespace, OverloadTracker>,
     subclasses_by_superclass: HashMap<QualifiedName, Vec<SubclassName>>,
-    has_unrepresentable_constructors: HashSet<QualifiedName>,
     nested_type_name_map: HashMap<QualifiedName, String>,
 }
 
@@ -198,7 +211,6 @@ impl<'a> FnAnalyzer<'a> {
             overload_trackers_by_mod: HashMap::new(),
             pod_safe_types: Self::build_pod_safe_type_set(&apis),
             subclasses_by_superclass: subclass::subclasses_by_superclass(&apis),
-            has_unrepresentable_constructors: HashSet::new(),
             nested_type_name_map: Self::build_nested_type_map(&apis),
         };
         let mut results = Vec::new();
@@ -321,7 +333,7 @@ impl<'a> FnAnalyzer<'a> {
         fun: Box<FuncToConvert>,
     ) -> Result<Box<dyn Iterator<Item = Api<FnPhase>>>, ConvertErrorWithContext> {
         let initial_name = name.clone();
-        let maybe_analysis_and_name = self.analyze_foreign_fn(name, &fun)?;
+        let maybe_analysis_and_name = self.analyze_foreign_fn(name, &fun);
 
         let (analysis, name) = match maybe_analysis_and_name {
             None => return Ok(Box::new(std::iter::empty())),
@@ -335,7 +347,7 @@ impl<'a> FnAnalyzer<'a> {
             FnKind::Method(sup, MethodKind::Constructor) => {
                 // Create a make_unique too
                 let make_unique_func = self.create_make_unique(&fun);
-                self.analyze_and_add_if_necessary(initial_name, make_unique_func, &mut results)?;
+                self.analyze_and_add_if_necessary(initial_name, make_unique_func, &mut results);
 
                 for sub in self.subclasses_by_superclass(sup) {
                     // Create a subclass constructor. This is a synthesized function
@@ -346,14 +358,14 @@ impl<'a> FnAnalyzer<'a> {
                         subclass_constructor_name.clone(),
                         subclass_constructor_func.clone(),
                         &mut results,
-                    )?;
+                    );
                     // and its corresponding make_unique
                     let make_unique_func = self.create_make_unique(&subclass_constructor_func);
                     self.analyze_and_add_if_necessary(
                         subclass_constructor_name,
                         make_unique_func,
                         &mut results,
-                    )?;
+                    );
                 }
             }
             FnKind::Method(
@@ -390,7 +402,7 @@ impl<'a> FnAnalyzer<'a> {
                     if !is_pure_virtual {
                         let maybe_wrap = create_subclass_fn_wrapper(sub, &super_fn_name, &fun);
                         let super_fn_name = ApiName::new_from_qualified_name(super_fn_name);
-                        self.analyze_and_add_if_necessary(super_fn_name, maybe_wrap, &mut results)?;
+                        self.analyze_and_add_if_necessary(super_fn_name, maybe_wrap, &mut results);
                     }
                 }
             }
@@ -412,8 +424,8 @@ impl<'a> FnAnalyzer<'a> {
         name: ApiName,
         new_func: Box<FuncToConvert>,
         results: &mut Vec<Api<FnPhase>>,
-    ) -> Result<(), ConvertErrorWithContext> {
-        let maybe_another_api = self.analyze_foreign_fn(name, &new_func)?;
+    ) {
+        let maybe_another_api = self.analyze_foreign_fn(name, &new_func);
         if let Some((analysis, name)) = maybe_another_api {
             results.push(Api::Function {
                 fun: new_func,
@@ -422,7 +434,6 @@ impl<'a> FnAnalyzer<'a> {
                 name_for_gc: None,
             });
         }
-        Ok(())
     }
 
     /// Take a constructor e.g. pub fn A_A(this: *mut root::A);
@@ -452,7 +463,7 @@ impl<'a> FnAnalyzer<'a> {
         &mut self,
         name: ApiName,
         fun: &FuncToConvert,
-    ) -> Result<Option<(FnAnalysis, ApiName)>, ConvertErrorWithContext> {
+    ) -> Option<(FnAnalysis, ApiName)> {
         let mut cpp_name = name.cpp_name_if_present().cloned();
         let ns = name.name.get_namespace();
 
@@ -460,7 +471,7 @@ impl<'a> FnAnalyzer<'a> {
         // because they influence our behavior in whether types can be used in
         // vectors, but that was all recorded in the POD analysis phase.
         if fun.is_deleted {
-            return Ok(None);
+            return None;
         }
 
         // Let's gather some pre-wisdom about the name of the function.
@@ -469,7 +480,7 @@ impl<'a> FnAnalyzer<'a> {
         // for diagnostics whilst we do that.
         let initial_rust_name = fun.ident.to_string();
         if initial_rust_name.ends_with("_destructor") {
-            return Ok(None);
+            return None;
         }
         let diagnostic_display_name = cpp_name.as_ref().unwrap_or(&initial_rust_name);
 
@@ -561,7 +572,7 @@ impl<'a> FnAnalyzer<'a> {
                 // virally as arguments on other allowlisted types. But we don't want
                 // to generate methods unless the user has specifically asked us to.
                 // It may, for instance, be a private type.
-                return Ok(None);
+                return None;
             }
 
             // Method or static method.
@@ -640,43 +651,72 @@ impl<'a> FnAnalyzer<'a> {
         // where the error occurred such that we can put a marker in the output
         // Rust code to indicate that a problem occurred (benefiting people using
         // rust-analyzer or similar). Make a closure to make this easy.
-        let contextualize_error = |err| ConvertErrorWithContext(err, Some(error_context.clone()));
+        let mut ignore_reason = Ok(());
+        let mut set_ignore_reason =
+            |err| ignore_reason = Err(ConvertErrorWithContext(err, Some(error_context.clone())));
 
         if matches!(kind, FnKind::Method(_, MethodKind::Constructor)) {
             // Annoyingly, in this case only, we need to convert the 'this' back to a pointer.
             // We will previously have treated it as a reference because all normal methods
             // take 'self' as a reference.
-            let (arg0, analysis0) = self
-                .convert_fn_arg(
-                    original_first_argument.unwrap(),
-                    ns,
-                    &rust_name,
-                    &fun.synthesized_this_type,
-                    &fun.references,
-                    false,
-                )
-                .map_err(contextualize_error)?;
-            params = std::iter::once(arg0)
-                .chain(params.into_iter().skip(1))
-                .collect();
-            param_details[0] = analysis0;
+            let r = self.convert_fn_arg(
+                original_first_argument.unwrap(),
+                ns,
+                &rust_name,
+                &fun.synthesized_this_type,
+                &fun.references,
+                false,
+            );
+            match r {
+                Ok((arg0, analysis0)) => {
+                    params = std::iter::once(arg0)
+                        .chain(params.into_iter().skip(1))
+                        .collect();
+                    param_details[0] = analysis0;
+                }
+                Err(e) => set_ignore_reason(e),
+            }
         }
 
         let requires_unsafe = self.should_be_unsafe(&param_details);
-        // Skip private methods; but if we've a private constructor, keep
-        // a note of it. We continue to process protected methods since,
-        // though they may not be callable elsewhere, we my be subclassing
-        // the class and need to handle them that way.
-        let generate_code = match fun.cpp_vis {
+
+        // Now we can add context to the error, check for a variety of error
+        // cases. In each case, we continue to record the API, because it might
+        // influence our later decisions to generate synthetic constructors
+        // or note whether the type is abstract.
+        let externally_callable = match fun.cpp_vis {
             CppVisibility::Private => {
-                if let FnKind::Method(self_ty, MethodKind::Constructor) = &kind {
-                    self.has_unrepresentable_constructors
-                        .insert(self_ty.clone());
-                }
-                return Ok(None);
+                set_ignore_reason(ConvertError::PrivateMethod);
+                false
             }
             CppVisibility::Protected => false,
             CppVisibility::Public => true,
+        };
+        if matches!(fun.special_member, Some(SpecialMemberKind::MoveConstructor)) {
+            set_ignore_reason(ConvertError::MoveConstructorUnsupported)
+        } else if fun.references.rvalue_ref_return {
+            set_ignore_reason(ConvertError::RValueReturn)
+        } else if !fun.references.rvalue_ref_params.is_empty() {
+            set_ignore_reason(ConvertError::RValueParam)
+        } else if let Some(problem) = bads.into_iter().next() {
+            match problem {
+                Ok(_) => panic!("No error in the error"),
+                Err(problem) => set_ignore_reason(problem),
+            }
+        } else if fun.unused_template_param {
+            // This indicates that bindgen essentially flaked out because templates
+            // were too complex.
+            set_ignore_reason(ConvertError::UnusedTemplateParam)
+        } else {
+            match kind {
+                FnKind::Method(_, MethodKind::Static) => {}
+                FnKind::Method(ref self_ty, _)
+                    if !known_types().is_cxx_acceptable_receiver(self_ty) =>
+                {
+                    set_ignore_reason(ConvertError::UnsupportedReceiver)
+                }
+                _ => {}
+            }
         };
 
         // The name we use within the cxx::bridge mod may be different
@@ -696,43 +736,6 @@ impl<'a> FnAnalyzer<'a> {
         }
         let mut cxxbridge_name = make_ident(&cxxbridge_name);
 
-        // Now we can add context to the error, check for a couple of error
-        // cases. First, see if any of the parameters are trouble.
-        if let Some(problem) = bads.into_iter().next() {
-            match problem {
-                Ok(_) => panic!("No error in the error"),
-                Err(problem) => return Err(contextualize_error(problem)),
-            }
-        }
-        // Second, reject any functions handling types which we flake out on.
-        if fun.unused_template_param {
-            return Err(contextualize_error(ConvertError::UnusedTemplateParam));
-        }
-        match kind {
-            FnKind::Method(_, MethodKind::Static) => {}
-            FnKind::Method(ref self_ty, _) => {
-                // Reject move constructors.
-                if matches!(fun.special_member, Some(SpecialMemberKind::MoveConstructor)) {
-                    self.has_unrepresentable_constructors
-                        .insert(self_ty.clone());
-                    return Err(contextualize_error(
-                        ConvertError::MoveConstructorUnsupported,
-                    ));
-                }
-                if !known_types().is_cxx_acceptable_receiver(self_ty) {
-                    return Err(contextualize_error(ConvertError::UnsupportedReceiver));
-                }
-            }
-            _ => {}
-        };
-        if fun.references.rvalue_ref_return {
-            return Err(contextualize_error(ConvertError::RValueReturn));
-        }
-        // Ensure we do this _after_ recording the presence of move constructors.
-        if !fun.references.rvalue_ref_params.is_empty() {
-            return Err(contextualize_error(ConvertError::RValueParam));
-        }
-
         // Analyze the return type, just as we previously did for the
         // parameters.
         let mut return_analysis = if let FnKind::Method(ref self_ty, MethodKind::MakeUnique) = kind
@@ -750,7 +753,10 @@ impl<'a> FnAnalyzer<'a> {
             }
         } else {
             self.convert_return_type(&fun.output, ns, &fun.references)
-                .map_err(contextualize_error)?
+                .unwrap_or_else(|err| {
+                    set_ignore_reason(err);
+                    ReturnTypeAnalysis::default()
+                })
         };
         let mut deps = params_deps;
         deps.extend(return_analysis.deps.drain());
@@ -759,9 +765,7 @@ impl<'a> FnAnalyzer<'a> {
         if num_input_references != 1 && return_analysis.was_reference {
             // cxx only allows functions to return a reference if they take exactly
             // one reference as a parameter. Let's see...
-            return Err(contextualize_error(ConvertError::NotOneInputReference(
-                rust_name,
-            )));
+            set_ignore_reason(ConvertError::NotOneInputReference(rust_name.clone()));
         }
         let mut ret_type = return_analysis.rt;
         let ret_type_conversion = return_analysis.conversion;
@@ -877,7 +881,7 @@ impl<'a> FnAnalyzer<'a> {
 
         // Naming, part two.
         // Work out our final naming strategy.
-        validate_ident_ok_for_cxx(&cxxbridge_name.to_string()).map_err(contextualize_error)?;
+        validate_ident_ok_for_cxx(&cxxbridge_name.to_string()).unwrap_or_else(set_ignore_reason);
         let rust_name_ident = make_ident(&rust_name);
         let (id, rust_rename_strategy) = match kind {
             FnKind::Method(..) | FnKind::TraitMethod { .. } => {
@@ -915,10 +919,11 @@ impl<'a> FnAnalyzer<'a> {
             vis,
             cpp_wrapper,
             deps,
-            generate_code,
+            ignore_reason,
+            externally_callable,
         };
         let name = ApiName::new_with_cpp_name(ns, id, cpp_name);
-        Ok(Some((analysis, name)))
+        Some((analysis, name))
     }
 
     /// Determine if this synthetic function should actually result in the implementation
@@ -1177,12 +1182,7 @@ impl<'a> FnAnalyzer<'a> {
         if self.config.exclude_impls {
             return;
         }
-        let types_without_constructors = Self::find_all_types(apis);
-        // For types with private constructors, we won't have generated code for them,
-        // but we equally don't want to synthesize a public constructor. The same applies
-        // to other types of constructor we might skip e.g. move constructors.
-        let mut types_without_constructors =
-            &types_without_constructors - &self.has_unrepresentable_constructors;
+        let mut types_without_constructors = Self::find_all_types(apis);
         // Now subtract all the actual constructors we know of.
         for api in apis.iter() {
             if let Api::Function {
@@ -1276,7 +1276,8 @@ impl Api<FnPhase> {
             Api::Function {
                 analysis: FnAnalysis {
                     cpp_wrapper: Some(..),
-                    generate_code: true,
+                    ignore_reason: Ok(_),
+                    externally_callable: true,
                     ..
                 },
                 ..

--- a/engine/src/conversion/codegen_cpp/mod.rs
+++ b/engine/src/conversion/codegen_cpp/mod.rs
@@ -152,7 +152,8 @@ impl<'a> CppCodeGenerator<'a> {
                     analysis:
                         FnAnalysis {
                             cpp_wrapper: Some(cpp_wrapper),
-                            generate_code: true,
+                            ignore_reason: Ok(_),
+                            externally_callable: true,
                             ..
                         },
                     fun,

--- a/engine/src/conversion/codegen_rs/fun_codegen.rs
+++ b/engine/src/conversion/codegen_rs/fun_codegen.rs
@@ -63,7 +63,7 @@ pub(super) fn gen_function(
     analysis: FnAnalysis,
     cpp_call_name: String,
 ) -> RsCodegenResult {
-    if !analysis.generate_code {
+    if analysis.ignore_reason.is_err() || !analysis.externally_callable {
         return RsCodegenResult::default();
     }
     let cxxbridge_name = analysis.cxxbridge_name;

--- a/engine/src/conversion/convert_error.rs
+++ b/engine/src/conversion/convert_error.rs
@@ -55,6 +55,7 @@ pub enum ConvertError {
     NonPublicNestedType,
     RValueParam,
     RValueReturn,
+    PrivateMethod,
 }
 
 fn format_maybe_identifier(id: &Option<Ident>) -> String {
@@ -102,6 +103,7 @@ impl Display for ConvertError {
             ConvertError::NonPublicNestedType => write!(f, "This type is nested within another struct/class with protected or private visibility.")?,
             ConvertError::RValueParam => write!(f, "This function takes an rvalue reference parameter (&&) which is not yet supported.")?,
             ConvertError::RValueReturn => write!(f, "This function returns an rvalue reference (&&) which is not yet supported.")?,
+            ConvertError::PrivateMethod => write!(f, "This method is private")?,
         }
         Ok(())
     }
@@ -132,6 +134,7 @@ impl std::fmt::Display for ErrorContext {
     }
 }
 
+#[derive(Clone)]
 pub(crate) struct ConvertErrorWithContext(pub(crate) ConvertError, pub(crate) Option<ErrorContext>);
 
 impl std::fmt::Debug for ConvertErrorWithContext {

--- a/engine/src/conversion/mod.rs
+++ b/engine/src/conversion/mod.rs
@@ -35,9 +35,14 @@ use crate::{CppCodegenOptions, CppFilePair, UnsafePolicy};
 
 use self::{
     analysis::{
-        abstract_types::mark_types_abstract, casts::add_casts, check_names, fun::FnPhase,
-        gc::filter_apis_by_following_edges_from_allowlist, pod::analyze_pod_apis,
-        remove_ignored::filter_apis_by_ignored_dependents, tdef::convert_typedef_targets,
+        abstract_types::{discard_ignored_functions, mark_types_abstract},
+        casts::add_casts,
+        check_names,
+        fun::FnPhase,
+        gc::filter_apis_by_following_edges_from_allowlist,
+        pod::analyze_pod_apis,
+        remove_ignored::filter_apis_by_ignored_dependents,
+        tdef::convert_typedef_targets,
     },
     api::{AnalysisPhase, Api},
     codegen_rs::RsCodeGenerator,
@@ -148,6 +153,8 @@ impl<'a> BridgeConverter<'a> {
                 Self::dump_apis_with_deps("analyze fns", &analyzed_apis);
                 let analyzed_apis = mark_types_abstract(self.config, analyzed_apis);
                 Self::dump_apis_with_deps("marking abstract", &analyzed_apis);
+                let analyzed_apis = discard_ignored_functions(analyzed_apis);
+                Self::dump_apis_with_deps("ignoring ignorable fns", &analyzed_apis);
                 // Remove any APIs whose names are not compatible with cxx.
                 let analyzed_apis = check_names(analyzed_apis);
                 // During parsing or subsequent processing we might have encountered
@@ -156,6 +163,7 @@ impl<'a> BridgeConverter<'a> {
                 // too.
                 let analyzed_apis = filter_apis_by_ignored_dependents(analyzed_apis);
                 Self::dump_apis_with_deps("removing ignored dependents", &analyzed_apis);
+
                 // We now garbage collect the ones we don't need...
                 let mut analyzed_apis =
                     filter_apis_by_following_edges_from_allowlist(analyzed_apis, self.config);

--- a/integration-tests/src/tests.rs
+++ b/integration-tests/src/tests.rs
@@ -7450,6 +7450,20 @@ fn test_abstract_up() {
 }
 
 #[test]
+fn test_abstract_private() {
+    let hdr = indoc! {"
+    #include <memory>
+    class A {
+        virtual void foo() const = 0;
+    public:
+        virtual ~A() {}
+    };
+    "};
+    let rs = quote! {};
+    run_test("", hdr, rs, &["A"], &[]);
+}
+
+#[test]
 fn test_class_having_protected_method() {
     let hdr = indoc! {"
     #include <cstdint>

